server/block/campfire.go

```diff
@@ -0,0 +1,280 @@
+type CampfireItem struct {
+	// Item is a specific item being cooked on top of the campfire.
+	Item item.Stack
+	// Time is the countdown of ticks until the food item is cooked (when 0).
+	Time time.Duration
+}
+
+// Model ...
+func (Campfire) Model() world.BlockModel {
+	return model.Campfire{}
+}
+
+// SideClosed ...
+func (Campfire) SideClosed(cube.Pos, cube.Pos, *world.World) bool {
+	return false
+}
+
+// BreakInfo ...
+func (c Campfire) BreakInfo() BreakInfo {
+	return newBreakInfo(2, alwaysHarvestable, axeEffective, func(t item.Tool, enchantments []item.Enchantment) []item.Stack {
+		var drops []item.Stack
+		if hasSilkTouch(enchantments) {
+			drops = append(drops, item.NewStack(c, 1))
+		} else {
+			switch c.Type {
+			case NormalFire():
+				drops = append(drops, item.NewStack(item.Charcoal{}, 2))
+			case SoulFire():
+				drops = append(drops, item.NewStack(SoulSoil{}, 1))
+			}
+		}
+		for _, v := range c.Items {
+			if !v.Item.Empty() {
+				drops = append(drops, v.Item)
+			}
+		}
+		return drops
+	})
+}
+
+// LightEmissionLevel ...
+func (c Campfire) LightEmissionLevel() uint8 {
+	if c.Extinguished {
+		return 0
+	}
+	return c.Type.LightLevel()
+}
+
+// Ignite ...
+func (c Campfire) Ignite(pos cube.Pos, w *world.World, _ world.Entity) bool {
+	w.PlaySound(pos.Vec3(), sound.Ignite{})
+	if !c.Extinguished {
+		return false
+	}
+	if _, ok := w.Liquid(pos); ok {
+		return false
+	}
+
+	c.Extinguished = false
+	w.SetBlock(pos, c, nil)
+	return true
+}
+
+// Splash ...
+func (c Campfire) Splash(w *world.World, pos cube.Pos, p potion.Potion) {
+	if p != potion.Water() || c.Extinguished {
+		// Water is the only potion that can extinguish a lit campfire.
+		return
+	}
+
+	w.PlaySound(pos.Vec3Centre(), sound.FireExtinguish{})
+	c.Extinguished = true
+	w.SetBlock(pos, c, nil)
+}
+
+// Activate ...
+func (c Campfire) Activate(pos cube.Pos, _ cube.Face, w *world.World, u item.User, ctx *item.UseContext) bool {
+	held, _ := u.HeldItems()
+	if held.Empty() {
+		return false
+	}
+
+	if _, ok := held.Item().(item.Shovel); ok && !c.Extinguished {
+		w.PlaySound(pos.Vec3Centre(), sound.FireExtinguish{})
+		c.Extinguished = true
+		w.SetBlock(pos, c, nil)
+
+		ctx.DamageItem(1)
+		return true
+	}
+
+	rawFood, ok := held.Item().(item.Smeltable)
+	if !ok || !rawFood.SmeltInfo().Food {
+		return false
+	}
+
+	for i, it := range c.Items {
+		if it.Item.Empty() {
+			c.Items[i] = CampfireItem{
+				Item: held.Grow(-held.Count() + 1),
+				Time: time.Second * 30,
+			}
+
+			ctx.SubtractFromCount(1)
+
+			w.PlaySound(pos.Vec3Centre(), sound.ItemAdd{})
+			w.SetBlock(pos, c, nil)
+			return true
+		}
+	}
+	return false
+}
+
+// UseOnBlock ...
+func (c Campfire) UseOnBlock(pos cube.Pos, face cube.Face, _ mgl64.Vec3, w *world.World, user item.User, ctx *item.UseContext) (used bool) {
+	if _, ok := w.Block(pos).(Campfire); ok && face == cube.FaceUp {
+		return false
+	}
+	pos, _, used = firstReplaceable(w, pos, face, c)
+	if !used {
+		return
+	}
```

Comment:
Looks like Bedrock Edition just generally does not let you place a campfire on top of another campfire, regardless of which face was tapped

```suggestion
	pos, _, used = firstReplaceable(w, pos, face, c)
	if !used {
		return
	}
	if _, ok := w.Block(pos.Side(cube.FaceDown)).(Campfire); ok {
		return false
	}
```